### PR TITLE
Fix DTO mapping when field not present in persistent entity

### LINKED ISSUE #1826: DTO projection property mapping incorrect
### Expected Behavior

When using DTO projection, I expect DTO projection to correctly map the property names to the SQL query, even if the properties are not included in the original entity. For example, if the class has an `authorName` field, I would expect this to be mapped to `author_name` in the ResultSet.

### Actual Behaviour

The wrong column name is being used to search the ResultSet for the value. In the linked example, the `authorName` is used to search the ResultSet, rather than `author_name`.

### Steps To Reproduce

1. Clone the attached repo
2. Run the application
3. Query the `/books/authors` endpoint

### Environment Information

- JDK version: 17

### Example Application

https://github.com/kieranjen/micronaut-dto-projection

### Version

3.7.3 but this started happening in 3.5.0 (it was working in 3.4.4)

### PR DESCRIPTION
**Pull request by radovanradic**
Fix DTO mapping when field not present in persistent entity, should fix #1826

### CODE REVIEW

File: data-jdbc/src/main/java/io/micronaut/data/jdbc/operations/DefaultJdbcRepositoryOperations.java [hunk 1]
```diff
@@ -324,11 +324,14 @@ public <T, R> R findOne(@NonNull PreparedQuery<T, R> pq) {
...
                         if (preparedQuery.isDtoProjection()) {
-                            TypeMapper<ResultSet, R> introspectedDataMapper = new DTOMapper<>(
-                                    persistentEntity,
-                                    columnNameResultSetReader,
-                                    jsonCodec,
-                                    conversionService);
+                            boolean isRawQuery = preparedQuery.getAnnotationMetadata().stringValue(Query.class, DataMethod.META_MEMBER_RAW_QUERY).isPresent();
+                            TypeMapper<ResultSet, R> introspectedDataMapper = new SqlDTOMapper<>(
+                                persistentEntity,
+                                isRawQuery ? getEntity(preparedQuery.getResultType()) : persistentEntity,
```
Comment (radovanradic) [inline data-jdbc/src/main/java/io/micronaut/data/jdbc/operations/DefaultJdbcRepositoryOperations.java:330]: Need to pass DTO entity so we can take persistent property if not found in mapped entity. This is how findAll works and findOne should work the same way.

File: data-runtime/src/main/java/io/micronaut/data/runtime/mapper/DTOMapper.java [hunk 2]
```diff
@@ -109,20 +109,27 @@ public Object read(@NonNull S object, @NonNull String name) throws ConversionErr
...
     }
 
     @Nullable
     @Override
     public Object read(@NonNull S object, @NonNull Argument<?> argument) {
-        RuntimePersistentProperty<T> pp = persistentEntity.getPropertyByName(argument.getName());
+        String name = argument.getName();
+        RuntimePersistentProperty<T> pp = persistentEntity.getPropertyByName(name);
         if (pp == null) {
+            if (persistentEntity != dtoEntity) {
```
Comment (radovanradic) [inline data-runtime/src/main/java/io/micronaut/data/runtime/mapper/DTOMapper.java:123]: Try to find persistent property in DTO entity if prop not found in mapped entity.

File: data-tck/src/main/java/io/micronaut/data/tck/entities/PatientDto.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    private final String name;
+
+    private final String history;
+
+    private final String doctorNotes;
+
+    public PatientDto(String name, String history, String doctorNotes) {
+        this.name = name;
+        this.history = history;
+        this.doctorNotes = doctorNotes;
```
Comment (radovanradic) [inline data-tck/src/main/java/io/micronaut/data/tck/entities/PatientDto.java:17]: Mapped entity Patient does not have doctorNotes field and DTO does have, tests with this entity were failing/issue reproducible before fix was checked in.

File: data-jdbc/src/main/java/io/micronaut/data/jdbc/operations/DefaultJdbcRepositoryOperations.java [hunk 4]
```diff
@@ -324,11 +324,14 @@ public <T, R> R findOne(@NonNull PreparedQuery<T, R> pq) {
                         return result;
                     } else if (rs.next()) {
                         if (preparedQuery.isDtoProjection()) {
-                            TypeMapper<ResultSet, R> introspectedDataMapper = new DTOMapper<>(
-                                    persistentEntity,
-                                    columnNameResultSetReader,
-                                    jsonCodec,
-                                    conversionService);
+                            boolean isRawQuery = preparedQuery.getAnnotationMetadata().stringValue(Query.class, DataMethod.META_MEMBER_RAW_QUERY).isPresent();
```
[the remark this answers is not in the document]
Comment (radovanradic) [reply data-jdbc/src/main/java/io/micronaut/data/jdbc/operations/DefaultJdbcRepositoryOperations.java]: Change in https://github.com/micronaut-projects/micronaut-data/pull/1828/commits/cd7239591d135aa7bee511431b65873f351b7403 (if I understood what you suggested)

**Review by graemerocher — APPROVED** [review]

**Comment by radovanradic** [comment]
Added fix for r2dbc and tests too.

**Review by dstepanov — APPROVED** [review]

**Pull request merged**